src/steadystateproblem.cpp

```diff
@@ -425,28 +432,35 @@ bool SteadystateProblem::checkConvergence(const Solver *solver,
                                           SensitivityMethod checkSensitivities) {
     /* get RHS and compute weighted error norm */
     if (checkSensitivities == SensitivityMethod::adjoint) {
-        getQBfromQ(model, xQ, xQB);
-        getQBfromQ(model, xB, xQBdot);
+        /* In the adjoitncase, only xQB contributes to the gradient,
+           the exact steadystate is less important, and xB = xQdot may even
+           not converge to zero. So we need xQBdot, hence compute xQB first. */
+        computeQBfromQ(model, xQ, xQB);
+        computeQBfromQ(model, xB, xQBdot);
         wrms = getWrmsNorm(xQB, xQBdot, solver->getAbsoluteToleranceQuadratures(),
                            solver->getRelativeToleranceQuadratures(), ewtQB);
-    } else {
-        model->fxdot(t, x, dx, xdot);
-        wrms = getWrmsNorm(x, xdot, solver->getAbsoluteToleranceSteadyState(),
-                           solver->getRelativeToleranceSteadyState(), ewt);
+        return wrms < RCONST(1.0);
     }
-    bool converged = wrms < RCONST(1.0);
 
-    /* get RHS of sensitivities and compute weighted error norm */
-    if (checkSensitivities == SensitivityMethod::forward) {
-        for (int ip = 0; ip < model->nplist(); ++ip) {
-            if (converged) {
-                sx = solver->getStateSensitivity(t);
-                model->fsxdot(t, x, dx, ip, sx[ip], dx, xdot);
-                wrms = getWrmsNorm(
-                    x, xdot, solver->getAbsoluteToleranceSteadyStateSensi(),
-                    solver->getRelativeToleranceSteadyStateSensi(), ewt);
-                converged = wrms < RCONST(1.0);
-            }
+    /* We're not in adjoint mode. Check convergence of xdot first */
+    model->fxdot(t, x, dx, xdot);
+    wrms = getWrmsNorm(x, xdot, solver->getAbsoluteToleranceSteadyState(),
+                       solver->getRelativeToleranceSteadyState(), ewt);
+
+    /* Do we need sensitivities? If not, we're done */
+    if (checkSensitivities == SensitivityMethod::none)
```

Comment:
```suggestion
    if (checkSensitivities != SensitivityMethod::forward)
```
We don't need convergence of `sx` for adjoint sensis, right?

Reply:
> If we're doing adjoint, we don't reach that point anyway but have returned before.

I assume this belongs here. Right its technically correct, but the code is not very explicit about the intentions here. This is also illustrated by the fact that the current code will break if adjoint is used. I would propose the following implementation:

```
if (checkSensitivities == SensitivityMethod::adjoint) {
        computeQBfromQ(model, xQ, xQB);
        computeQBfromQ(model, xB, xQBdot);
        wrms = getWrmsNorm(xQB, xQBdot, solver->getAbsoluteToleranceQuadratures(),
                           solver->getRelativeToleranceQuadratures(), ewtQB);
} else {
        model->fxdot(t, x, dx, xdot);
        wrms = getWrmsNorm(x, xdot, solver->getAbsoluteToleranceSteadyState(),
                           solver->getRelativeToleranceSteadyState(), ewt);
}
bool converged = wrms < RCONST(1.0);

if (checkSensitivities != SensitivityMethod::forward)
        return converged;
```

Does that make sense?

Reply:
It's logically equivalent, so totally fine to do that... Will implement (addressing the code review atm anyway)

Reply:
http://nemo.oudeis.fyi/posts/make-it-explicit/

To clarify why I make a fuss about this: Try writing down sentences for every line that says exactly what the code is supposed to be doing including all conditionals (if this becomes overly lengthy, its a good indication that the code is too convoluted). If your comment says

> /* Do we need sensitivities? If not, we're done */

but the code actually implements

> if (checkSensitivities == SensitivityMethod::none)

that raises a red flag with me.

Reply:
> http://nemo.oudeis.fyi/posts/make-it-explicit/
> 
> To clarify why I make a fuss about this: Try writing down sentences for every line that says exactly what the code is supposed to be doing including all conditionals (if this becomes overly lengthy, its a good indication that the code is too convoluted). If your comment says
> 
> > /* Do we need sensitivities? If not, we're done */
> 
> but the code actually implements
> 
> > if (checkSensitivities == SensitivityMethod::none)
> 
> that raises a red flag with me.

No, that's totally okay to make a fuzz about it. I now start to actually see your point and not only your comment.
Maybe on a more general note: Are there tutorials about software development you can recommend?A substantial amount of our working time goes into this, but I have never attended a lecture or a course about this, it's all learning by doing. If there's an easy way to improve the situation (good free online course//tutorial) that's highly welcome...

Reply:
> > http://nemo.oudeis.fyi/posts/make-it-explicit/
> > To clarify why I make a fuss about this: Try writing down sentences for every line that says exactly what the code is supposed to be doing including all conditionals (if this becomes overly lengthy, its a good indication that the code is too convoluted). If your comment says
> > > /* Do we need sensitivities? If not, we're done */
> > 
> > 
> > but the code actually implements
> > > if (checkSensitivities == SensitivityMethod::none)
> > 
> > 
> > that raises a red flag with me.
> 
> No, that's totally okay to make a fuzz about it. I now start to actually see your point and not only your comment.
> Maybe on a more general note: Are there tutorials about software development you can recommend?A substantial amount of our working time goes into this, but I have never attended a lecture or a course about this, it's all learning by doing. If there's an easy way to improve the situation (good free online course//tutorial) that's highly welcome...

Unfortunately no. Usually everybody has their own opinion and it's often language specific. @dweindl might have some pointers.

Reply:
As you may see from the post, giving good advice is also not easy ;)

Reply:
software development is a pretty wide field. depends on what you are looking for specifically. maybe we discuss that elsewhere.